### PR TITLE
[BSC] Ellipsis

### PR DESCRIPTION
**Pull request by Shelvak**

### CODE REVIEW

File: contracts/ControllerStratAbs.sol [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        if (fee > 0) { want.safeTransfer(treasury, fee); }
+    }
+
+    function wantBalance() public view returns (uint) {
+        return want.balanceOf(address(this));
+    }
+
+    function balance() public view returns (uint) {
+        return wantBalance() + balanceOfPoolInWant();
+    }
```
Comment (batudo) [inline contracts/ControllerStratAbs.sol:211]:
```suggestion
    }

```

File: contracts/ControllerStratAbs.sol [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    function balanceOfPool() public view virtual returns (uint) {
+        // should be implemented
+    }
```
Comment (batudo) [inline contracts/ControllerStratAbs.sol:214]:
```suggestion
    }

```

File: contracts/ControllerEllipsisStrat.sol [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        if (_amount > 0) {
+            return POOL.calc_withdraw_one_coin(_amount, TOKEN_INDEX);
+        } else {
+            return 0;
+        }
+    }
+    function balanceOfPool() public view override returns (uint) {
+        (uint _amount, ) = STAKE.userInfo(STAKE_POOL_ID, address(this));
+        return _amount;
+    }
```
Comment (batudo) [inline contracts/ControllerEllipsisStrat.sol:205]:
```suggestion
    }

```

File: contracts/ControllerEllipsisStrat.sol [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        if (wantToPoolToken > poolTokenAmount) { poolTokenAmount = wantToPoolToken; }
+    }
+
+    function calc_withdraw_one_coin(uint _amount) public view returns (uint) {
+        if (_amount > 0) {
+            return POOL.calc_withdraw_one_coin(_amount, TOKEN_INDEX);
+        } else {
+            return 0;
+        }
+    }
```
Comment (batudo) [inline contracts/ControllerEllipsisStrat.sol:201]:
```suggestion
    }

```

File: contracts/ControllerEllipsisStrat.sol [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "hardhat/console.sol";
+
+import "./ControllerStratAbs.sol";
+import "../interfaces/IEps.sol";
```
Comment (batudo) [inline contracts/ControllerEllipsisStrat.sol:11]:
```suggestion
import "../interfaces/IEps.sol";
```
This file would be added later?

File: contracts/ControllerEllipsisStrat.sol [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        poolTokenAmount = poolTokenAmount * (RATIO_PRECISION - poolSlippageRatio) / RATIO_PRECISION;
+
+        // In case the pool is unbalanced (attack), make a double check for
+        // the expected amount with minExpected set ratios.
+        uint wantToPoolToken = _minWantToPoolToken(_amount);
+
+        if (wantToPoolToken > poolTokenAmount) { poolTokenAmount = wantToPoolToken; }
+    }
+
+    function calc_withdraw_one_coin(uint _amount) public view returns (uint) {
```
Comment (batudo) [inline contracts/ControllerEllipsisStrat.sol:195]:
```suggestion
    function calcWithdrawOneCoin(uint _amount) public view returns (uint) {
```

File: contracts/ControllerEllipsisStrat.sol (hunk 5 above)
Comment (Shelvak) [reply contracts/ControllerEllipsisStrat.sol:11]: Good catch :+1:

**Pull request merged**